fem/fe/fe_rbf.hpp

```diff
@@ -0,0 +1,556 @@
+                                  };
+      return ident[Check(rbfType)];
+   }
+
+   // Convert identifier to rbf int
+   static int GetType(const char rbfIdent)
+   {
+      switch (rbfIdent)
+      {
+         case 'G': return Gaussian;
+         case 'M': return Multiquadric;
+         case 'I': return InvMultiquadric;
+         case 'T': return TruncatedGaussian;
+         case 'C': return CompactGaussian;
+         case '1': return Wendland11;
+         case '3': return Wendland31;
+         case '6': return Wendland33;
+      }
+      MFEM_ABORT("unknown RBF identifier: " << rbfIdent);
+      return -1;
+   }
+};
+
+// Dimensionless distance metrics, whose output is the input for RBFs
+class DistanceMetric
+{
+protected:
+   int dim;
+public:
+   DistanceMetric(int D) { dim = D; }
+   virtual ~DistanceMetric() { }
+
+   virtual void SetDim(int D) { dim = D; }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const = 0;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const = 0;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const = 0;
+
+   static DistanceMetric *GetDistance(int dim, int pnorm);
+};
+
+// Dimensionless distance with r = |x| + |y| + ...
+class L1Distance : public DistanceMetric
+{
+public:
+   L1Distance(int D) : DistanceMetric(D) { };
+   virtual ~L1Distance() { }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const;
+};
+
+// Dimensionless distance with r = (x^2 + y^2 + ...)^(1/2)
+class L2Distance : public DistanceMetric
+{
+public:
+   L2Distance(int D) : DistanceMetric(D) { };
+   virtual ~L2Distance() { }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const;
+};
+
+// Dimensionless distance with r = (x^p + y^p + ...)^(1/p)
+class LpDistance : public DistanceMetric
+{
+   const int p;
+   const double pinv;
+public:
+   LpDistance(int D, int pnorm)
+      : DistanceMetric(D),
+        p(pnorm),
+        pinv(1. / static_cast<double>(p))
+   { };
+   virtual ~LpDistance() { }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const;
+};
+
+
+/* Pure virtual class for a finite element with radial basis functions
+   instead of polynomials inside each element */
+class KernelFiniteElement : public ScalarFiniteElement
```

Comment:
Why the word `kernel`?

Reply:
Since reproducing kernels aren't a radial basis function (their value depends not only on position but on the value of other kernels at the same evaluation point), I went with the common thread that both RBF and RK use kernels commonly used in meshless methods instead of polynomials. If we can think of another name that fits better, I would be happy to change it. We could call it RKFiniteElement if we wanted to, since that will often be what is used, but this class also includes RBF elements.